domain/application/service/application.go

```diff
@@ -190,42 +206,84 @@ func makeAddUnitArgs(in AddUnitArg) application.AddUnitArg {
 // returning an error satisfying [applicationerrors.ApplicationNotFoundError]
-// if the application doesn't exist.
-func (s *ApplicationService) UpsertCAASUnit(ctx context.Context, appName string, unit UpsertCAASUnitParams) error {
+// if the application doesn't exist. If the application life is Dead, an error
+// satisfying [applicationerrors.ApplicationIsDead] is returned.
+func (s *ApplicationService) RegisterCAASUnit(ctx context.Context, appName string, args RegisterCAASUnitParams) error {
+	if args.PasswordHash == nil {
+		return errors.NotValidf("password hash")
+	}
+	if args.ProviderId == nil {
+		return errors.NotValidf("provider id")
+	}
+	if !args.OrderedScale {
+		return errors.NewNotImplemented(nil, "registering CAAS units not supported without ordered unit IDs")
+	}
+	if args.UnitName == "" {
+		return errors.NotValidf("missing unit name")
+	}
+
 	p := AddUnitArg{
-		UnitName:     &unit.UnitName,
-		PasswordHash: unit.PasswordHash,
+		UnitName:     &args.UnitName,
+		PasswordHash: args.PasswordHash,
+		CloudContainer: &CloudContainerParams{
+			ProviderId: args.ProviderId,
+			Ports:      args.Ports,
+		},
 	}
-	if unit.ProviderId != nil || unit.Address != nil || unit.Ports != nil {
-		cldContainer := &CloudContainerParams{
-			ProviderId: unit.ProviderId,
-			Ports:      unit.Ports,
+	if args.Address != nil {
+		addr := network.NewSpaceAddress(*args.Address, network.WithScope(network.ScopeMachineLocal))
+		// k8s doesn't support spaces yet.
+		addr.SpaceID = network.AlphaSpaceId
+		p.CloudContainer.Address = &addr
+		origin := network.OriginProvider
+		p.CloudContainer.AddressOrigin = &origin
+	}
+	// We need to do a bunch of business logic in the one transaction so pass in a closure that is
+	// given the transaction to use.
+	unitArg := makeUpsertUnitArgs(p)
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		appID, err := s.st.GetApplicationID(ctx, appName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		unitLife, err := s.st.UnitLife(ctx, args.UnitName)
+		if errors.Is(err, uniterrors.NotFound) {
+			return s.insertCAASUnit(ctx, appID, args.OrderedId, unitArg)
 		}
-		if unit.Address != nil {
-			addr := network.NewSpaceAddress(*unit.Address, network.WithScope(network.ScopeMachineLocal))
-			// k8s doesn't support spaces yet.
-			addr.SpaceID = network.AlphaSpaceId
-			cldContainer.Address = &addr
-			origin := network.OriginProvider
-			cldContainer.AddressOrigin = &origin
+		if unitLife == life.Dead {
+			return fmt.Errorf("dead unit %q already exists%w", args.UnitName, errors.Hide(applicationerrors.ApplicationIsDead))
 		}
-		p.CloudContainer = cldContainer
+		return s.st.UpsertUnit(ctx, appID, unitArg)
+	})
+	return errors.Annotatef(err, "saving caas unit %q", args.UnitName)
+}
+
+func (s *ApplicationService) insertCAASUnit(
+	ctx context.Context, appID coreapplication.ID, orderedID int, args application.UpsertUnitArg,
```

Comment:
What's the background behind "ordered" in this context?

Reply:
It's a k8s thing - refers to stateful set semantics around pod spin up / tear down. Is direct port of mongo logic.